akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpEntity.scala

```diff
@@ -91,6 +91,10 @@ sealed trait HttpEntity extends jm.HttpEntity {
    * of its streaming nature. If the dataBytes source is materialized a second time, it will fail with an
    * "stream can cannot be materialized more than once" exception.
    *
+   * When called on `Strict` entities, the above warnings can be ignored. Discarding is not required because
+   * the underlying [[akka.stream.scaladsl.Source]] will have already been drained.
```

Comment:
I’d propose to reword to be mor precise, if something was already drained that is also the same situation as before. This is about that it is “already completely available in memory”, so there is no need to materialize the stream, it just has the strict data readily available

Reply:
Wait, can you expand on that? So if the underlying source has already been read, discarding won't throw the exception, even if the entity is not `Strict`.
That makes sense. But now I am not so sure what exactly determines whether the exception is thrown.

`Source.empty` passed into an `HttpEntity.apply` won't do it.
`Source.single` would be your above example.
Even if I pass in a `Source` from an infinite `Stream`, I don't seem to see this exception, instead the discarding will just go on forever :)

Sorry but I am still missing a vital piece of understanding :(

Reply:
>  So if the underlying source has already been read, discarding won't throw the exception, even if the entity is not Strict. That makes sense. 

No, that's a case where it *would* throw. The thing is: if it's a real stream, you can materialize it once, because if you had to materialize it twice but expect "to get the data from the beginning of the stream" that means you're forced to buffer the entire thing in memory - otherwise you would not be able to send the first bytes, someone has already consumed them. This is the reason the double materialization is an issue – either you always have to buffer things, and be ready for multiple materializations or you don't buffer, and multiple ones would not get the same data so we prefer to just fail rather than send partial entity bytes. The same applies for any real streaming server, not just akka.

We have something in the works that will buffer "until a given point" (memory size), but again, it's the same trade off of real streaming vs. you wanting to materialize many times, which means it can't be completely fully streaming (since the data source could be a live tcp connection etc).

> Source.empty passed into an HttpEntity.apply won't do it.

This is special, "empty" in general is special most of the time ;)

> Source.single would be your above example.

Again, here the thing is in memory and is just a normal source, this would materialize multiple times just fine. A REAL source from a REAL tcp connection would not.

> Even if I pass in a Source from an infinite Stream, I don't seem to see this exception, instead the discarding will just go on forever :)

Again, you pass them in as "normal" Sources, they can be run as many times as you want. A "Source" that is connected to something "live" would not be able to.

Reply:
Hope this explains it somewhat :)

Reply:
Ah, I think I get it now. Thank you very much. So, if I wanted to whip up a unit test against my own code to make sure a function properly discards entities, I would actually need to connect to some sort of "outside" `Source` rather than quickly whipping up one as a sort of "mock object".

I will update the wording and add to the Java DSL some time today.

Reply:
Yeah, exactly - it's the property of that "source of data" that introduces this issue here.

Thanks a ton! :)